Archivist.lua

```diff
@@ -68,39 +67,146 @@ function Archivist:Warn(valid, pattern, ...) -- Like assert, but doesn't interru
 	end
 end
 
-function Archivist:IsInitialized()
-	return self.initialized
+local serializeConfig = {
+	errorOnUnserializableType = false
+}
+
+-- prototype used to create archives
+local proto = {}
+Archivist.proto = proto
+
+--@debug@
+proto.debug = true
+--@end-debug@
+
+proto.Assert = Archivist.Assert
+proto.Warn = Archivist.Warn
+
+do -- function Archive:GenerateID()
+	-- adapted from https://gist.github.com/jrus/3197011
+	local function randomHex()
+		return ('%x'):format(math.random(0, 0xf))
+	end
+
+	function proto:GenerateID()
+		local template ='xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx'
+		return (template:gsub('x', randomHex))
+	end
+end
+
+function Archivist:IsInitialized(sv)
+	return self.archives[sv] ~= nil
 end
 
--- Give Archivist its archive to play with. Called automatically, unless Archivist has been embedded.
-function Archivist:Initialize(sv)
+-- Create an archivist instance. Called automatically, unless Archivist has been embedded.
+-- If called with an already initialized savedvariables table, then the previously initialized instance is returned.
+function Archivist:Initialize(sv, prototypes)
 	do -- arg validation
-		self:Assert(not self:IsInitialized(), "Archivist has already been initialized.")
-		self:Assert(type(sv) == "table", "Attempt to initialize Archivist SavedVariables with a %q instead of a table.", type(sv))
+		self:Assert(type(sv) == "table", "Invalid argument #1 to Initialize, expected table but got %q.", type(sv))
+		self:Assert(type(sv.internalVersion) ~= "number" or sv.internalVersion <= self.internalVersion, "Invalid argument #1 to Initialize, savedvariable is encoded using a newer version of Archivist. Please upgrade this installation if possible.")
+		self:Assert(prototypes == nil or type(prototypes) == "table", "Invalid argument #2 to Initialize, expected table or nil but got %q", type(prototypes))
+	end
+	if self:IsInitialized(sv) then
+		return self.archives[sv]
+	end
+	if type(sv.stores) ~= "table" then
+		sv.stores = {}
+	end
+	local archive = setmetatable({
+		prototypes = {},
+		activeStores = {},
+		storeMap = {},
+		sv = sv,
+		initialized = true,
+	}, {
+		__index = proto
+	})
+	if type(sv.internalVersion) ~= "number" or sv.internalVersion < self.internalVersion then
```

Comment:
I've got a slight concern here that doing migrations upon initialization of the instance _may_ cause script timeout errors for people with gargantuan amounts of data, considering that this first migration will be touching every single store to decompress, deserialize, re-serialize, and re-compress things.

Even if it isn't a significant issue now, any future migrations which do similar things would potentially have a stacking effect in the case where multiple migrations need to be performed.

It might be better to apply this migration on a store-by-store basis at the point when an individual store is opened by checking the internal versioning at a store level to see which deserializer should be used to initially read it.

Reply:
This is a good point. I think I'll work on this some more in a follow-on change, once rossnichols/LibSerialize#4 & SafeteeWoW/LibDeflate#6 are complete.